File: 0587-erect-the-fence/0587-erect-the-fence.py
class Solution:
    def is_left(self, p1: List[int], p2: List[int], m: List[int]):
        # (piM x pi pj).uz >= 0 if m to the right of segment pipj
        # x is vectorial product
        return ((p2[0]-p1[0])*(m[1]-p1[1])-(p2[1]-p1[1])*(m[0]-p1[0])) > 0

    def outerTrees(self, trees: List[List[int]]) -> List[List[int]]:
        hull = []
        trees.sort(key=lambda x: (x[0], x[1]))  # left to right, down to up if equality

        # lower hull: left to right
        for tree in trees:
            while (len(hull) > 1) and self.is_left(hull[-1], hull[-2], tree):
                hull.pop()
            hull.append(tuple(tree))
        hull.pop()  # remove last duplicated element

        # upper hull: left to right
        for tree in trees[::-1]:
            while (len(hull) > 1) and self.is_left(hull[-1], hull[-2], tree):
                hull.pop()
            hull.append(tuple(tree))  # remove last duplicated element

        # remove duplcated element
        return list(set(hull))

# fail to implement
"""
from scipy.spatial import ConvexHull
class Solution:
    def outerTrees(self, trees: List[List[int]]) -> List[List[int]]:
        hull = ConvexHull(points=trees)
        trees_on_perimeter = set()
        for p in hull.vertices:
            trees_on_perimeter.add(tuple(trees[p]))
        return [list(t) for t in trees_on_perimeter]
"""
# TLE, O(n^3) hull algo complexity
"""
class Solution:
    def is_left(self, p1: List[int], p2: List[int], m: List[int]):
        # (piM x pi pj).uz >= 0 if m to the right of segment pipj
        # x is vectorial product
        return ((p2[0]-p1[0])*(m[1]-p1[1])-(p2[1]-p1[1])*(m[0]-p1[0]))>0
    
    def convex_hull(self, points: List[List[int]]) -> List[Tuple[Tuple[int,int], Tuple[int,int]]]:
        N = len(points)
        if N == 1:
            return [(tuple(points[0]),tuple(points[0]))]
        hull = []
        for i in range(N):
            for j in range(N):
                if i != j:
                    border = True
                    for k in range(N):
                        if (k != i) and (k != j):
                            if self.is_left(points[i], points[j], points[k]):
                                border = False
                                break
                    if border:
                        hull.append((tuple(points[i]),tuple(points[j])))
        return hull

    def outerTrees(self, trees: List[List[int]]) -> List[List[int]]:
        hull = self.convex_hull(trees)
        points_on_perimeter = set()
        for p1, p2 in hull:
            points_on_perimeter.add(p1)
            points_on_perimeter.add(p2)
        return [list(p) for p in points_on_perimeter]
"""

File: 0587-erect-the-fence/0587-erect-the-fence.rs
use std::collections::HashSet;
impl Solution {
    pub fn is_left(p1: (i32,i32), p2: (i32,i32), m: (i32,i32)) -> bool {
        // (piM x pi pj).uz >= 0 if m to the right of segment pipj
        // x is vectorial product
        return ((p2.0-p1.0)*(m.1-p1.1)-(p2.1-p1.1)*(m.0-p1.0)) > 0;
    }
    pub fn outer_trees(mut trees: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        if (trees.len() == 1) {
            return trees;
        }
        let mut hull: Vec<(i32, i32)> = Vec::new();
        trees.sort_unstable_by_key(|item| (item[0], item[1]));;  // left to right, down to up if equality

        // lower hull: left to right
        for t in 0..trees.len() {
            let tree: (i32,i32) = (trees[t][0], trees[t][1]);
            while (hull.len() > 1) {
                if (Solution::is_left(hull[(hull.len() as usize)-1], hull[(hull.len() as usize)-2], tree)) {
                    hull.pop();
                }
                else {
                    break;
                }
            }
            hull.push(tree);
        }
        hull.pop();  // remove last duplicated element

        // upper hull: left to right
        for t in (0..trees.len()).rev() {
            let tree: (i32,i32) = (trees[t][0], trees[t][1]);
            while (hull.len() > 1) {
                if (Solution::is_left(hull[(hull.len() as usize)-1], hull[(hull.len() as usize)-2], tree)) {
                    hull.pop();
                }
                else {
                    break;
                }
            }
            hull.push(tree);
        }
        hull.pop();  // remove last duplicated element

        // remove duplcated element
        let mut res: Vec<Vec<i32>> = Vec::new();
        let mut set: HashSet<(i32,i32)> = HashSet::new();
        for i in 0..hull.len() {
            let tree: (i32,i32) = hull[i];
            if (!set.contains(&tree)) {
                set.insert(tree);
                let mut to_add: Vec<i32> = Vec::new();
                to_add.push(tree.0);
                to_add.push(tree.1);
                res.push(to_add);
            }
        }
        return res;
    }
}